### DISCUSSION TITLE
Swift bindings

### DISCUSSION THREAD

**Discussion by Stefterv** [google/filament#5747] [Ideas]
I have started writing bindings for Swift over at https://github.com/Stefterv/filament/tree/swift_bindings 
Is this somethings I could contribute to the project or is it out of scope for now? Like mentioned in #4111

The current state is that I have got the repo working as a Swift Package.
I have added an early attempt at writing the bindings themselves. They were written with my specific project in mind so they are in no way complete yet, the plan is to improve from here.

**Comment by prideout** [comment]
Thanks, this is awesome, but it will need to be in a fork or separate project.  Maintaining our current set of language bindings is already difficult since our team is so small, this would be out of scope for now.

**Reply by romainguy** [reply]
Unless it can be generated... :D

**Reply by Stefterv** [reply]
Okay the bindings I will create/maintain within a fork then. 

What about distributing Filament through the Swift Package Manager, is that something that could be contributed? This would mean it becomes possible to add Filament to a project with just the Github url. This is only a one time change to the project and wouldn't need much active maintenance

**Reply by bejado** [reply]
I'm not familiar with the Swift Package Manager, but could the Filament bindings live under the Stefterv/filament GitHub URL?

**Reply by Stefterv** [reply]
To make clear, the creation of a SwiftPM Package does not require the creation of any bindings at all, it will only contain the compiled C++ code and its headers. It's an alternative to the Cocaopods system. 

The Swift Bindings would be in their own repo, not as a fork of the main filament repo, e.g. Stefterv/FilamentSwift.  My bindings repository will require the published package coming from the main repo and adds the bindings on top of the compiled C++ code.
I am asking to integrate the SwiftPM setup into the main repo as it requires modification to the CMakeFiles and other buildscripts. 

If these changes happen the situation will be as follows: when a new Filament version is published, the build systems could automatically provide a new version of Filament as a Swift Package.  In most cases, when a new version is published the main Filament Package will be updated instantly and the bindings will stay on the old version, until I release updated bindings for the new version.

PS. I finished the main push for the bindings today and I've got the gltf-viewer example working to the point where Filament is running, just need to load the assets tomorrow.

**Reply by Stefterv** [reply]
I can write/have written the majority of the updates that need to be made to the build/publishing system, I will just need some guidance how to do it neatly as I'm not familiar enough with CMake

**Reply by bejado** [reply]
It's not clear to me why our CMake files would need to be modified. I'm reluctant to do this because anything that goes into our repo we need to support, and as @prideout mentioned we're a small team and our release process is already complex.

It sounds like the bindings would be separate from our project, but even supporting Filament _without bindings_ as a SwiftPM Package might be too involved for us.

**Reply by Stefterv** [reply]
The changes that need to be made are in this commit: https://github.com/Stefterv/filament/commit/47e7c36dde0efd80563fd5f78a948e9613cd2ad9. It involves generating `.framework` first and secondly `.xcframework` files alongside the `.a` files as the output of the cmake process, resulting in the `out/ios-release/filament/lib/universal` folder. SwiftPM doesn't support the inclusion of `.a` binary targets as Cocoapods does, so I had to create an adaption for that. 
I suspect it can even be improved to the point where it doesn't need a giant list of the `.binaryTargets` in the `Package.swift` meaning even if the libraries/targets within Filament change there needs to be no change made to the Swift Package.

I am also advocating for this change as the main part of the project as it will make including Filament into any iOS project much easier (in XCode click file->Add Package... and paste the url) which I think is neat. 

As for the effort it would take to maintain it, I took a look at the commit history of the .podspec file and apart from version bumps, it has only been changed 5 times since its creation. I suspect the SwiftPM setup to very similar or even less work. I'm happy to do the initial bulk of the work of setting it up. 

I do understand that with any systemic change to a build process there is always the risk of unintended side-effects, the uptick is though that the iOS release process of Filament might be easier and more future proof according to this [article](https://anuragajwani.medium.com/distributing-universal-ios-frameworks-as-xcframeworks-using-cocoapods-699c70a5c961)

**Reply by Stefterv** [reply]
I revisited the bindings this weekend and I found a way to generate the necessary files without modifying the original repo and finished most of the bindings: https://github.com/Stefterv/filament_swift the example should work apart from the skybox bug mentioned in #7151

**Reply by Stefterv** [reply]
Oh and almost forgot to mention [Mixing Swift and C++](https://www.swift.org/documentation/cxx-interop/) Swift/Apple are working on direct C++ interoperability, I also played around with it but most of the Filament classes are not compatible and it's still in pretty early stages so I decided to finish the Swift -> ObjC -> C++ bindings
